monix-eval/jvm/src/test/scala/monix/eval/TaskLocalJVMSuite.scala

```diff
@@ -174,4 +177,87 @@ object TaskLocalJVMSuite extends SimpleTestSuite {
       assertEquals(r, 0)
     }
   }
+
+  testAsync("local state is encapsulated by Task run loop on single thread") {
+    implicit val s = TracingScheduler(Scheduler.singleThread("local-test"))
+      .withExecutionModel(ExecutionModel.AlwaysAsyncExecution)
+    implicit val opts = Task.defaultOptions.enableLocalContextPropagation
+
+    def runAssertion(run: Task[Unit] => Any, method: String): Future[Unit] = {
+      val p = Promise[Unit]
+      val local = Local(0)
+      val task = Task.evalAsync(local := 50).guarantee(Task(p.success(())).void)
+
+      run(task)
+
+      // Future could still carry isolated local
+      // because it was created inside isolated block
+      val f = Local.isolate(p.future)
```

Comment:
Does `isolate` actually do anything here?

Reply:
Yes, otherwise it will carry the state of `Local` from the place of `Promise` completion. Not sure if it's a bug or a feature